test/spmd/tensor/parallel/test_tp_examples.py

```diff
@@ -0,0 +1,376 @@
+                module.register_parameter("bias", replicated_bias)
+
+    def replicate_input(inputs):
+        return DTensor.from_local(inputs[0], device_mesh, replicate)
+
+    def aggregate_output(outputs):
+        assert isinstance(outputs, DTensor)
+        return (
+            outputs.redistribute(outputs.device_mesh, replicate)
+            .contiguous()
+            .to_local()
+        )
+
+    dist_mod = distribute_module(
+        m,
+        device_mesh,
+        partition_fn=shard_params,
+        input_fn=replicate_input,
+        output_fn=aggregate_output,
+    )
+    return dist_mod
+
+
+def shard_self_attn(m, device_type, tp_size):
+    start_idx = 0
+    device_mesh = DeviceMesh(
+        device_type,
+        list(range(start_idx, start_idx + tp_size)),
+    )
+    col_wise_sharding = [Shard(0)]
+    row_wise_sharding = [Shard(1)]
+    replicate = [Replicate()]
+
+    def shard_params(name, module):
+        if isinstance(module, nn.Linear):
+            if name == "qkv":
+                sharded_weight = nn.Parameter(
+                    distribute_tensor(
+                        module.weight, device_mesh, col_wise_sharding
+                    )
+                )
+                module.register_parameter("weight", sharded_weight)
+                module.weight.register_hook(
+                    functools.partial(_gradient_hook, module.weight)
+                )
+                if module.bias is not None:
+                    sharded_bias = nn.Parameter(
+                        distribute_tensor(
+                            module.bias, device_mesh, col_wise_sharding
+                        )
+                    )
+                    module.register_parameter("bias", sharded_bias)
+                    module.bias.register_hook(
+                        functools.partial(_gradient_hook, module.bias)
+                    )
+            elif name == "proj":
+                sharded_weight = nn.Parameter(
+                    distribute_tensor(
+                        module.weight, device_mesh, row_wise_sharding
+                    )
+                )
+                module.register_parameter("weight", sharded_weight)
+                module.weight.register_hook(
+                    functools.partial(_gradient_hook, module.weight)
+                )
+                _aggregate_local_tensor(module)
+                if module.bias is not None:
+                    replicated_bias = nn.Parameter(
+                        distribute_tensor(module.bias, device_mesh, replicate)
+                    )
+                    module.register_parameter("bias", replicated_bias)
+                    module.bias.register_hook(
+                        functools.partial(_gradient_hook, module.bias)
+                    )
+
+    def replicate_input(inputs):
+        DTensors = []
+        for tensor in inputs:
+            DTensors.append(DTensor(tensor, device_mesh, replicate))
```

Comment:
why we can using `DTensor` constructor here instead of `from_local`? Let's not use DTensor constructor directly in non-op impls, as we probably need to make the __new__ constructor be more restrictive and only be used internally (we need to add things like shape, stride to the constructor as it will be hard to use by caller)

Reply:
Sure. Let's use the public API.